File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/token_accept/TokenAcceptViewModel.kt
@file:Suppress("PackageNaming", "LongMethod", "TooManyFunctions", "MagicNumber")

package app.pantopus.android.ui.screens.token_accept

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.token_accept.GuestPassDto
import app.pantopus.android.data.api.models.token_accept.HomeInviteDetailsDto
import app.pantopus.android.data.api.models.token_accept.HomeInviteResponse
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.auth.AuthRepository
import app.pantopus.android.data.homes.HomeInvitationPreview
import app.pantopus.android.data.homes.homeTaskUUID
import app.pantopus.android.data.token_accept.TokenAcceptRepository
import app.pantopus.android.ui.screens.homes.claim_review.HomeClaimSessionScope
import app.pantopus.android.ui.theme.PantopusIcon
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.Duration
import java.time.Instant
import javax.inject.Inject
import kotlin.math.ceil

/**
 * Resolves an invite token into one of three offers, then drives
 * accept / decline via the matching route. Mirrors iOS exactly.
 */
@HiltViewModel
class TokenAcceptViewModel
    @Inject
    constructor(
        private val repository: TokenAcceptRepository,
        private val auth: AuthRepository,
        savedStateHandle: SavedStateHandle,
        private val invitations: HomeInvitationDecisionFactory,
    ) : ViewModel() {
        private val token: String = savedStateHandle.get<String>(TOKEN_KEY) ?: ""

        private val _state = MutableStateFlow<TokenAcceptUiState>(TokenAcceptUiState.Loading)
        val state: StateFlow<TokenAcceptUiState> = _state.asStateFlow()

        /** Increments when the user finishes the screen so the host
         *  can pop the back stack / dismiss. */
        private val _dismissEvents = MutableStateFlow(0)
        val dismissEvents: StateFlow<Int> = _dismissEvents.asStateFlow()

        private var generation = 0L
        private var visible = false
        private var lifetime: Job? = null
        private var session: HomeClaimSessionScope? = null

        fun pause() {
            generation++
            visible = false
            lifetime?.cancel()
            lifetime = null
            session = null
            _state.value = TokenAcceptUiState.Loading
        }

        fun load() {
            pause()
            visible = true
            val revision = generation
            val job = SupervisorJob(viewModelScope.coroutineContext[Job])
            lifetime = job
            val scope = CoroutineScope(viewModelScope.coroutineContext + job)
            val current = invitations.session(scope)
            session = current
            if (token.isBlank() || token.length > 512) {
                _state.value = TokenAcceptUiState.Expired("Missing or invalid invitation link.")
                return
            }
            scope.launch {
                current.invalidated.collect { invalidated ->
                    if (invalidated && visible && generation == revision) {
                        pause()
                        _state.value = TokenAcceptUiState.Error("Your session changed. Reopen the invitation to continue.")
                    }
                }
            }
            scope.launch {
                try {
                    current.requireCurrent()
                    val original =
                        try {
                            invitations.hasOriginal(current)
                        } catch (
                            cancelled: CancellationException,
                        ) {
                            throw cancelled
                        } catch (_: Exception) {
                            true
                        }
                    current.requireCurrent()
                    if (!isCurrent(revision)) return@launch
                    if (original) {
                        _state.value = TokenAcceptUiState.HomeInvitation
                        return@launch
                    }
                    val resolved = resolveOffer()
                    current.requireCurrent()
                    if (isCurrent(revision)) _state.value = resolved
                } catch (cancelled: CancellationException) {
                    throw cancelled
                } catch (_: Exception) {
                    if (visible && generation == revision) {
                        _state.value = TokenAcceptUiState.Error("The invitation could not be checked right now. Try again.")
                    }
                }
            }
        }

        private suspend fun resolveOffer(): TokenAcceptUiState =
            coroutineScope {
                val homeAsync = async { invitations.preview(token) }
                val seatAsync = async { repository.businessSeatInvite(token) }
                val guestAsync = async { repository.guestPass(token) }
                val home = homeAsync.await()
                val seat = seatAsync.await()
                val guest = guestAsync.await()
                when {
                    home == HomeInvitationPreview.Found -> TokenAcceptUiState.HomeInvitation
                    seat is NetworkResult.Success && homeTaskUUID(seat.data.seatId) ->
                        TokenAcceptUiState.Ready(makeSeatOffer(seat.data, identityChip()))
                    guest is NetworkResult.Success && guest.data.pass != null ->
                        TokenAcceptUiState.Ready(makeGuestOffer(guest.data.pass, identityChip()))
                    home == HomeInvitationPreview.Missing && missing(seat) && missing(guest) ->
                        TokenAcceptUiState.Expired("This invitation could not be found. Check the complete link with the sender.")
                    else -> TokenAcceptUiState.Error("The invitation could not be checked right now. Try again.")
                }
            }

        private fun isCurrent(revision: Long): Boolean = visible && generation == revision && session?.isCurrent == true

        private fun missing(result: NetworkResult<*>): Boolean = result is NetworkResult.Failure && result.error.code == 404

        fun accept() {
            val ready = _state.value as? TokenAcceptUiState.Ready ?: return
            val revision = generation
            if (!isCurrent(revision)) return
            val offer = ready.offer
            if (offer.inviteType == InviteType.HomeInvite) {
                _state.value = TokenAcceptUiState.HomeInvitation
                return
            }
            _state.value = TokenAcceptUiState.Accepting(offer)
            viewModelScope.launch {
                val result =
                    when (offer.inviteType) {
                        InviteType.BusinessSeat -> repository.acceptBusinessSeat(token)
                        else -> NetworkResult.Success(Unit)
                    }
                if (!isCurrent(revision)) return@launch
                _state.value =
                    when (result) {
                        is NetworkResult.Success<*> ->
                            TokenAcceptUiState.Accepted(
                                offer,
                                if (offer.inviteType == InviteType.GuestPass) {
                                    "Your guest pass is ready to view."
                                } else {
                                    "Your business seat acceptance is saved."
                                },
                            )
                        is NetworkResult.Failure ->
                            TokenAcceptUiState.Error(
                                "The acceptance could not be confirmed. Retry to check the invitation's current status.",
                            )
                    }
            }
        }

        fun decline() {
            val ready = _state.value as? TokenAcceptUiState.Ready ?: return
            val revision = generation
            if (!isCurrent(revision)) return
            val offer = ready.offer
            if (offer.inviteType == InviteType.HomeInvite) {
                _state.value = TokenAcceptUiState.HomeInvitation
                return
            }
            _state.value = TokenAcceptUiState.Accepting(offer)
            viewModelScope.launch {
                val result =
                    if (offer.inviteType == InviteType.BusinessSeat) {
                        repository.declineBusinessSeat(
                            token,
                        )
                    } else {
                        NetworkResult.Success(Unit)
                    }
                if (!isCurrent(revision)) return@launch
                _state.value =
                    when (result) {
                        is NetworkResult.Success<*> -> TokenAcceptUiState.Declined
                        is NetworkResult.Failure ->
                            TokenAcceptUiState.Error(
                                "The decline could not be confirmed. Retry to check the invitation's current status.",
                            )
                    }
            }
        }

        fun dismiss() {
            pause()
            _dismissEvents.value = _dismissEvents.value + 1
        }

        // MARK: - Projection

        private fun identityChip(): IdentityChipContent {
            val signed = auth.state.value as? AuthRepository.State.SignedIn
            val user = signed?.user
            val label = user?.displayName?.takeIf { it.isNotEmpty() } ?: user?.email ?: "Accepting as guest"
            return IdentityChipContent(label = label)
        }

        companion object {
            const val TOKEN_KEY = "token"

            internal fun makeHomeOffer(
                home: HomeInviteResponse,
                invitation: HomeInviteDetailsDto,
                identity: IdentityChipContent,
            ): TokenAcceptOffer {
                val homeName = home.home?.name ?: "this home"
                val city = home.home?.city
                val venue =
                    listOfNotNull(homeName.takeIf { it.isNotBlank() }, city?.takeIf { it.isNotBlank() })
                        .joinToString(" · ")
                val sender = home.inviter?.name ?: home.inviter?.username ?: "Someone"
                val role = humanRole(invitation.proposedRole ?: "member")
                return TokenAcceptOffer(
                    invitationId = invitation.id,
                    inviteType = InviteType.HomeInvite,
                    title = "Join a home",
                    sender = "$sender invited you",
                    roleOffered = role,
                    venue = venue.ifEmpty { homeName },
                    benefits = homeBenefits(invitation.proposedRole),
                    expiry = formatExpiry(invitation.expiresAt),
                    safetyBand =
                        SafetyBand(
                            icon = PantopusIcon.Lock,
                            text = "Your email and personal account stay private — $sender only sees your accepted role.",
                        ),
                    primaryCtaLabel = "Join $homeName",
                    secondaryCtaLabel = "Decline",
                    identityChip = identity,
                )
            }

            internal fun makeSeatOffer(
                seat: app.pantopus.android.data.api.models.token_accept.BusinessSeatInviteResponse,
                identity: IdentityChipContent,
            ): TokenAcceptOffer {
                val venue = seat.business?.name ?: seat.business?.username ?: "this business"
                val sender = seat.business?.name ?: "The team"
                val role = humanRole(seat.roleBase ?: "member")
                return TokenAcceptOffer(
                    invitationId = seat.seatId,
                    inviteType = InviteType.BusinessSeat,
                    title = "Accept a business seat",
                    sender = "$sender offered you a seat",
                    roleOffered = role,
                    venue = venue,
                    benefits = seatBenefits(seat.roleBase),
                    expiry = null,
                    safetyBand =
                        SafetyBand(
                            icon = PantopusIcon.ShieldCheck,
                            text = "Your seat is firewalled — coworkers see your business profile, not your local identity.",
                        ),
                    primaryCtaLabel = "Add me to $venue",
                    secondaryCtaLabel = "Decline",
                    identityChip = identity,
                )
            }

            internal fun makeGuestOffer(
                pass: GuestPassDto,
                identity: IdentityChipContent,
            ): TokenAcceptOffer {
                val venue = pass.homeName ?: pass.customTitle ?: "the host's place"
                val kind = (pass.kind ?: "guest").replace('_', ' ')
                val label = pass.label ?: pass.customTitle ?: "Guest pass"
                return TokenAcceptOffer(
                    invitationId = null,
                    inviteType = InviteType.GuestPass,
                    title = label,
                    sender = "Welcome to $venue",
                    roleOffered = humanRole(kind),
                    venue = venue,
                    benefits = guestBenefits(pass.welcomeMessage, pass.expiresAt),
                    expiry = formatExpiry(pass.expiresAt),
                    safetyBand =
                        SafetyBand(
                            icon = PantopusIcon.Lock,
                            text = "Guest passes never reveal your account email — you stay anonymous to the host.",
                        ),
                    primaryCtaLabel = "View guest pass",
                    secondaryCtaLabel = "Not now",
                    identityChip = identity,
                )
            }

            internal fun humanRole(raw: String): String {
                val normalized = raw.replace('_', ' ').trim()
                if (normalized.isEmpty()) return "Member"
                return normalized.first().uppercaseChar() + normalized.drop(1)
            }

            internal fun homeBenefits(role: String?): List<String> {
                val lower = role?.lowercase().orEmpty()
                return when {
                    lower.contains("owner") || lower.contains("co_owner") ->
                        listOf(
                            "Co-manage occupants, ownership, and home settings",
                            "Share home docs, wi-fi, and entry info with guests",
                            "See all home activity in your Hub",
                        )
                    lower.contains("renter") || lower.contains("tenant") ->
                        listOf(
                            "See house docs, wi-fi, and entry info",
                            "Get notified about home updates and tasks",
                            "Mark yourself as a resident in your local profile",
                        )
                    else ->
                        listOf(
                            "See house docs, wi-fi, and entry info",
                            "Get home updates in your Hub",
                            "Privately label yourself a resident if you want",
                        )
                }
            }

            internal fun seatBenefits(role: String?): List<String> {
                val lower = role?.lowercase().orEmpty()
                val list =
                    mutableListOf<String>(
                        "Post and respond as ${humanRole(role ?: "member")}",
                        "Access the business dashboard and team feed",
                    )
                if (lower.contains("admin") || lower.contains("manager")) {
                    list.add("Invite teammates and manage seats")
                } else {
                    list.add("Switch identities anytime in the You tab")
                }
                return list
            }

            internal fun guestBenefits(
                welcomeMessage: String?,
                expiresAt: String?,
            ): List<String> {
                val list = mutableListOf<String>()
                welcomeMessage?.takeIf { it.isNotBlank() }?.let(list::add)
                list.add("See wi-fi, parking, and entry info during your stay")
                daysFromNow(expiresAt)?.let { days ->
                    list.add("Valid for $days day${if (days == 1) "" else "s"}")
                }
                return list
            }

            internal fun formatExpiry(iso: String?): String? {
                if (iso.isNullOrBlank()) return null
                val instant = runCatching { Instant.parse(iso) }.getOrNull() ?: return null
                val zoned = instant.atZone(java.time.ZoneId.systemDefault())
                val formatter = java.time.format.DateTimeFormatter.ofPattern("MMM d, yyyy h:mm a")
                return "Expires ${zoned.format(formatter)}"
            }

            internal fun daysFromNow(iso: String?): Int? {
                if (iso.isNullOrBlank()) return null
                val instant = runCatching { Instant.parse(iso) }.getOrNull() ?: return null
                val seconds = Duration.between(Instant.now(), instant).seconds
                if (seconds <= 0) return null
                return ceil(seconds.toDouble() / 86400.0).toInt()
            }
        }
    }

File: frontend/apps/ios/Pantopus/Features/TokenAccept/TokenAcceptViewModel.swift
//
//  TokenAcceptViewModel.swift
//  Pantopus
//
//  Resolves Home, business-seat and guest-pass links without treating
//  unavailable reads as expired invitations. Home decisions recover their
//  protected original before a newly opened token can replace the screen.
//

import Foundation
import Observation

@Observable
@MainActor
public final class TokenAcceptViewModel {
    public private(set) var state: TokenAcceptState = .loading
    private(set) var homeDecision: HomeInvitationDecisionViewModel?
    private var generation = 0
    private var session: HomeClaimSessionScope?
    var sessionIsCurrent: Bool {
        session?.isCurrent ?? true
    }

    func suspend() {
        generation += 1
        homeDecision?.suspend()
        homeDecision = nil
        leasePreview = nil
        state = .loading
    }

    func close() {
        suspend()
        onDeclined()
    }

    private let api: APIClient
    private let token: String
    private let leaseInvitation: Bool
    private var leasePreview: LeaseInvitePreviewResponse?
    private let auth: AuthManager
    private let invitationStore: any PendingHomeInvitationDecisionStoring
    private let onAccepted: @MainActor (InviteType) -> Void
    private let onDeclined: @MainActor () -> Void

    init(
        token: String,
        leaseInvitation: Bool = false,
        api: APIClient = .shared,
        auth: AuthManager? = nil,
        invitationStore: any PendingHomeInvitationDecisionStoring = PendingHomeInvitationDecisionStore(),
        onAccepted: @escaping @MainActor (InviteType) -> Void = { _ in },
        onDeclined: @escaping @MainActor () -> Void = {}
    ) {
        self.token = token
        self.leaseInvitation = leaseInvitation
        self.api = api
        self.auth = auth ?? api.authProvider ?? AuthManager.shared
        self.invitationStore = invitationStore
        self.onAccepted = onAccepted
        self.onDeclined = onDeclined
    }

    public func load() async {
        suspend()
        let revision = generation
        let lease = HomeClaimSessionScope(api: api)
        session = lease
        let scope = HomeInvitationDecisionViewModel.scope(api: api)
        guard scope.isValid, lease.isCurrent else {
            state = .error(message: "Sign in to review this invitation.")
            return
        }
        do {
            if try invitationStore.load(scope: scope) != nil {
                homeDecision = HomeInvitationDecisionViewModel(token: token, api: api, store: invitationStore)
                return
            }
        } catch {
            // An unreadable protected slot must not be treated as empty.
            homeDecision = HomeInvitationDecisionViewModel(token: token, api: api, store: invitationStore)
            return
        }
        if leaseInvitation {
            await loadLease(revision: revision, session: lease)
            return
        }
        let identity = identityChip()
        async let homeTask = resolve(JSONValue.self, endpoint: TokenAcceptEndpoints.homeInvite(token: token))
        async let seatTask = resolve(BusinessSeatInviteResponse.self, endpoint: TokenAcceptEndpoints.businessSeatInvite(token: token))
        async let guestTask = resolve(TokenAcceptGuestPassResponse.self, endpoint: TokenAcceptEndpoints.guestPass(token: token))
        let (home, seat, guest) = await (homeTask, seatTask, guestTask)
        guard generation == revision, lease.isCurrent, !Task.isCancelled else { return }
        if case let .success(value) = home {
            guard HomeInvitationValidation.preview(value) else {
                state = .error(message: "The invitation response could not be checked. Retry to review its current details.")
                return
            }
            homeDecision = HomeInvitationDecisionViewModel(token: token, api: api, store: invitationStore)
            return
        }
        if case let .success(value) = seat, let id = value.seatId, HomePostalValidation.uuid(id) {
            state = .ready(Self.makeSeatOffer(seat: value, identity: identity))
            return
        }
        if case let .success(value) = guest, let pass = value.pass, pass.label != nil || pass.customTitle != nil {
            state = .ready(Self.makeGuestOffer(pass: pass, identity: identity))
            return
        }
        let allMissing = home.isMissing && seat.isMissing && guest.isMissing
        state = allMissing ? .expired(message: "This invitation is unavailable. Check the complete link with the sender.")
            : .error(message: "The invitation could not be checked right now. Retry to review its current status.")
    }

    public func accept() async {
        guard case let .ready(offer) = state, sessionIsCurrent else { return }
        let revision = generation
        state = .accepting(offer)
        do {
            switch offer.inviteType {
            case .homeInvite:
                homeDecision = HomeInvitationDecisionViewModel(token: token, api: api, store: invitationStore)
            case .leaseInvite:
                guard let preview = leasePreview, case let .signedIn(user) = auth.state else { throw APIError.invalidResponse }
                let response: LeaseInviteAcceptanceResponse = try await api.request(TokenAcceptEndpoints.acceptLeaseInvite(token: token))
                guard generation == revision, sessionIsCurrent, !Task.isCancelled else { return }
                guard response.lease.homeId == preview.home.id, response.lease.primaryResidentUserId == user.id,
                      response.lease.state == "active", HomePostalValidation.uuid(response.lease.id),
                      response.occupancy.homeId == preview.home.id, response.occupancy.userId == user.id,
                      HomePostalValidation.uuid(response.occupancy.id), response.occupancy.isActive,
                      response.occupancy.verificationStatus == "verified" else { throw APIError.invalidResponse }
                state = .accepted(offer, message: "Your lease acceptance is saved.")
            case .businessSeat:
                let _: BusinessSeatAcceptResponse = try await api.request(
                    TokenAcceptEndpoints.acceptBusinessSeat(body: BusinessSeatAcceptBody(token: token))
                )
                guard generation == revision, sessionIsCurrent, !Task.isCancelled else { return }
                state = .accepted(offer, message: "Welcome to \(offer.venue) — your seat is active.")
                onAccepted(.businessSeat)
            case .guestPass:
                // Guest passes don't have a separate "accept" — viewing
                // is the acceptance. Mark accepted so the success
                // frame renders, then route the user to the pass.
                state = .accepted(offer, message: "Your guest pass is active. Welcome to \(offer.venue).")
                onAccepted(.guestPass)
            }
        } catch {
            guard generation == revision, sessionIsCurrent else { return }
            state = .error(message: "Acceptance could not be confirmed. Retry to check the invitation’s current status.")
        }
    }

    public func decline() async {
        guard case let .ready(offer) = state, sessionIsCurrent else { return }
        if offer.inviteType == .leaseInvite {
            suspend()
            onDeclined()
            return
        }
        let revision = generation
        state = .accepting(offer)
        do {
            switch offer.inviteType {
            case .leaseInvite: return
            case .homeInvite:
                homeDecision = HomeInvitationDecisionViewModel(token: token, api: api, store: invitationStore)
                return
            case .businessSeat:
                let _: AnyDecodable = try await api.request(
                    TokenAcceptEndpoints.declineBusinessSeat(body: BusinessSeatDeclineBody(token: token))
                )
            case .guestPass:
                break
            }
            guard generation == revision, sessionIsCurrent, !Task.isCancelled else { return }
            state = .declined
            onDeclined()
        } catch {
            guard generation == revision, sessionIsCurrent else { return }
            state = .error(message: "The decline could not be confirmed. Retry to check the invitation's current status.")
        }
    }

    // MARK: - Projection

    static func makeSeatOffer(
        seat: BusinessSeatInviteResponse,
        identity: IdentityChipContent
    ) -> TokenAcceptOffer {
        let venue = seat.business?.name ?? seat.business?.username ?? "this business"
        let sender = seat.business?.name ?? "The team"
        let role = humanRole(seat.roleBase ?? "member")
        return TokenAcceptOffer(
            invitationId: seat.seatId,
            inviteType: .businessSeat,
            title: "Accept a business seat",
            sender: "\(sender) offered you a seat",
            roleOffered: role,
            venue: venue,
            benefits: seatBenefits(role: seat.roleBase),
            expiry: nil,
            safetyBand: SafetyBand(
                icon: .shieldCheck,
                text: "Your seat is firewalled — coworkers see your business profile, not your local identity."
            ),
            primaryCtaLabel: "Add me to \(venue)",
            secondaryCtaLabel: "Decline",
            identityChip: identity
        )
    }

    static func makeGuestOffer(
        pass: TokenAcceptGuestPassDTO,
        identity: IdentityChipContent
    ) -> TokenAcceptOffer {
        let venue = pass.homeName ?? pass.customTitle ?? "the host's place"
        let kind = (pass.kind ?? "guest").replacingOccurrences(of: "_", with: " ")
        let label = pass.label ?? pass.customTitle ?? "Guest pass"
        let expiry = formatExpiry(pass.expiresAt)
        return TokenAcceptOffer(
            invitationId: nil,
            inviteType: .guestPass,
            title: label,
            sender: "Welcome to \(venue)",
            roleOffered: humanRole(kind),
            venue: venue,
            benefits: guestBenefits(welcomeMessage: pass.welcomeMessage, expiresAt: pass.expiresAt),
            expiry: expiry,
            safetyBand: SafetyBand(
                icon: .lock,
                text: "Guest passes never reveal your account email — you stay anonymous to the host."
            ),
            primaryCtaLabel: "View guest pass",
            secondaryCtaLabel: "Not now",
            identityChip: identity
        )
    }

    // MARK: - Helpers

    private func identityChip() -> IdentityChipContent {
        if case let .signedIn(user) = auth.state {
            let displayName = user.displayName ?? ""
            let label = displayName.isEmpty ? user.email : displayName
            return IdentityChipContent(label: label, handle: nil)
        }
        return IdentityChipContent(label: "Accepting as guest")
    }

    private enum Resolution<Value> {
        case success(Value), missing, unavailable
        var isMissing: Bool {
            if case .missing = self { true } else { false }
        }
    }

    private func resolve<T: Decodable>(_: T.Type, endpoint: Endpoint) async -> Resolution<T> {
        do { return try await .success(api.request(endpoint)) } catch APIError.notFound { return .missing } catch { return .unavailable }
    }

    static func humanRole(_ raw: String) -> String {
        let normalized = raw.replacingOccurrences(of: "_", with: " ").trimmingCharacters(in: .whitespaces)
        guard !normalized.isEmpty else { return "Member" }
        return normalized.prefix(1).uppercased() + normalized.dropFirst()
    }

    static func seatBenefits(role: String?) -> [String] {
        let lower = role?.lowercased() ?? ""
        var benefits: [String] = [
            "Post and respond as \(humanRole(role ?? "member"))",
            "Access the business dashboard and team feed"
        ]
        if lower.contains("admin") || lower.contains("manager") {
            benefits.append("Invite teammates and manage seats")
        } else {
            benefits.append("Switch identities anytime in the You tab")
        }
        return benefits
    }

    static func guestBenefits(welcomeMessage: String?, expiresAt: String?) -> [String] {
        var benefits: [String] = []
        if let message = welcomeMessage, !message.isEmpty {
            benefits.append(message)
        }
        benefits.append("See wi-fi, parking, and entry info during your stay")
        if let expiry = expiresAt, let days = daysFromNow(expiry) {
            benefits.append("Valid for \(days) day\(days == 1 ? "" : "s")")
        }
        return benefits
    }

    static func formatExpiry(_ iso: String?) -> String? {
        guard let iso, let date = ISO8601DateFormatter().date(from: iso) else { return nil }
        let formatter = DateFormatter()
        formatter.dateStyle = .medium
        formatter.timeStyle = .short
        return "Expires \(formatter.string(from: date))"
    }

    static func daysFromNow(_ iso: String) -> Int? {
        guard let date = ISO8601DateFormatter().date(from: iso) else { return nil }
        let seconds = date.timeIntervalSinceNow
        if seconds <= 0 { return nil }
        return Int((seconds / 86400).rounded(.up))
    }
}

extension TokenAcceptViewModel {
    private func loadLease(revision: Int, session: HomeClaimSessionScope) async {
        do {
            guard token.range(of: "^[a-fA-F0-9]{64}$", options: .regularExpression) != nil else { throw APIError.invalidResponse }
            let preview: LeaseInvitePreviewResponse = try await api.request(TokenAcceptEndpoints.leaseInvite(token: token))
            guard generation == revision, session.isCurrent, !Task.isCancelled else { return }
            guard case let .signedIn(user) = auth.state,
                  preview.accountEmail.lowercased() == user.email.lowercased(), HomePostalValidation.uuid(preview.home.id),
                  ["pending", "accepted"].contains(preview.invitation.status),
                  let start = HomeInvitationValidation.date(preview.invitation.proposedStart),
                  let expiry = HomeInvitationValidation.date(preview.invitation.expiresAt),
                  preview.invitation.status == "accepted" || expiry > Date() else { throw APIError.invalidResponse }
            if let end = preview.invitation.proposedEnd {
                guard let date = HomeInvitationValidation.date(end), date > start else { throw APIError.invalidResponse }
            }
            leasePreview = preview
            state = .ready(Self.makeLeaseOffer(preview))
        } catch {
            guard generation == revision, session.isCurrent, !Task.isCancelled else { return }
            leasePreview = nil
            if case APIError.clientError(status: 410, message: _) = error {
                state = .expired(message: "This lease invitation is closed or expired. Ask the landlord for a new invitation.")
            } else {
                state = .error(message: "This lease invitation could not be checked. Use the account it was sent to and retry.")
            }
        }
    }

    static func makeLeaseOffer(_ preview: LeaseInvitePreviewResponse) -> TokenAcceptOffer {
        let invitation = preview.invitation
        var dates = ["Starts: \(invitation.proposedStart.prefix(10))"]
        if let end = invitation.proposedEnd { dates.append("Ends: \(end.prefix(10))") }
        return TokenAcceptOffer(
            invitationId: nil,
            inviteType: .leaseInvite,
            title: "Review your lease invitation",
            sender: "Your landlord invited you",
            roleOffered: "Tenant",
            venue: [preview.home.name ?? "Your rental Home", preview.home.city].compactMap { $0 }.joined(separator: ", "),
            benefits: dates,
            expiry: formatExpiry(invitation.expiresAt),
            safetyBand: SafetyBand(
                icon: .shieldCheck,
                text: "Review the Home and dates before accepting. Access is checked again when you accept."
            ),
            primaryCtaLabel: invitation.status == "accepted" ? "Check saved acceptance" : "Accept lease invitation",
            secondaryCtaLabel: "Not now",
            identityChip: IdentityChipContent(label: preview.accountEmail)
        )
    }
}

/// Decodable placeholder used when we don't care about a response
/// body (e.g. the decline endpoints).
private struct AnyDecodable: Decodable {}
